### DISCUSSION TITLE
Metrics Publisher Support

### DISCUSSION THREAD

**Discussion by RodrasSilva** [aws/aws-sdk-kotlin#1241] [General]
Hello. 

I'm migrating my application from the Java AWS SDK to the Kotlin AWS SDK. 
When I was building the S3Client using the Java SDK,  I was able to provide a Metrics publisher where I could build my logic to transform the SDK-provided metrics into my own:
```kotlin 
 return S3AsyncClient.builder()
     .overrideConfiguration { cfg ->
          cfg.addMetricPublisher(S3MetricsPublisher(...))
      }  
     ...
```
```kotlin 
internal class S3MetricsPublisher(
    private val s3Metrics: ImageRepositoryMetrics,
    private val bucket: String
) : MetricPublisher {

    override fun publish(metricCollection: MetricCollection) {
        val metrics = metricsToMap(metricCollection)
        metrics[RETRY_COUNT_KEY.key]?.let { s3Metrics.publishRetryCount(bucket, (it as Int).toDouble()) }
        ...
    } 
```

How am I able to replicate this behavior using the Kotlin SDK and the telemetry provider?

Thanks in advance.

**Comment by ianbotsf** [comment]
Hi @RodrasSilva, excellent question! The telemetry and metrics interfaces for the AWS SDK for Kotlin are slightly different from the AWS SDK for Java but it still should be possible to implement your S3 publisher adaptor. The key is providing your own implementation of a `MetricsProvider` which receives calls to create/record metrics and then adapts them to your publishing logic. A possible implementation can be broken down into several parts:

### Providers

As an example, here's an implementation of a telemetry provider which has a custom meter provider and nothing else:

```kotlin
class MyTelemetryProvider : TelemetryProvider {
    override val contextManager = ContextManager.None
    override val loggerProvider = LoggerProvider.None
    override val tracerProvider = TracerProvider.None
    override val meterProvider = MyMeterProvider()
}

class MyMeterProvider : MeterProvider {
    override fun getOrCreateMeter(scope: String) = MyMeter(scope)
}

class MyMeter : Meter {
    …
}
```

### Meters

The `Meter` class supports creating instruments for 7 different types of measurements:
* **async up-down counter**: used for discrete measurements which go up and down like HTTP connection usage
* **monotonic counter**: used for discrete measurements which only go up like the number of calls made
* **`double` histogram**: used for aggregating non-discrete measurements which are likely to be statistically useful like latency
* **`long` histogram**: currently unused
* **`double` gauge**: currently unused
* **`long` gauge**: currently unused
* **up-down counter**: currently unused

Your code for a `Meter` implementation will vary depending on which metrics you want to capture and publish. You may choose to implement handling for every kind of instrument or only for one or two based on your use case. Here is an example of implementing support for capturing `double` histograms which would include measurements like latency:

```kotlin
class MyMeter : Meter {
    override fun createDoubleHistogram(
        name: String,
        units: String?,
        description: String?,
    ) = MyDoubleHistogram(name, units, description)

    // Override the other Meter members with no-op implementations
    override fun createAsyncUpDownCounter(…) = NoOpAsyncMeasurementHandle
    override fun createUpDownCounter(…) = NoOpUpDownCounter
    override fun createMonotonicCounter(…) = NoOpMonotonicCounter
    override fun createLongHistogram(…) = NoOpLongHistogram
    override fun createLongGauge(…) = NoOpAsyncMeasurementHandle
    override fun createDoubleGauge(…) = NoOpAsyncMeasurementHandle
}

class MyDoubleHistogram(private val name: String, private val units: String?) : DoubleHistogram {
    override fun record(value: Double, attributes: Attributes, context: Context?) {
        // Handle the measurement
    }
}
```

### Asynchronous batching

The `record` function will be called synchronously for every measurement emitted. Several measurements will be emitted at different times during a given operation invocation so it's highly recommended to publish them asynchronously. For example, you may create a publishing adapter class which uses a [`BlockingQueue`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html) to collect measurements synchronously and a periodic function to drain the queue and publish them:

```kotlin
data class Measurement(val name: String, val value: Double, val attributes: Attributes)

class MyS3Publisher(private val s3: S3Client, private val coroutineScope: CoroutineScope) {
    private val queue = LinkedBlockingQueue<Measurement>()

    init {
        // Launch a background coroutine to publish metrics every 5 seconds
        coroutineScope.launch {
            while (isActive) {
                delay(5.seconds)
                publish()
            }
        }
    }

    fun put(measurement: Measurement) {
        queue.add(measurement)
    }

    private suspend fun publish() {
        val metrics = mutableListOf<Measurement>()
            .also { queue.drainTo(it) }
            .joinToString(separator = "\n")

        s3.putObject {
            bucket = "foo"
            key = "bar"
            body = ByteStream.fromString(metrics)
        }
    }
}
```

### Wiring it all together

Now that we have a publisher that can asynchronously batch requests, we just need to modify the providers and meters to accept an instance:

```kotlin
class MyTelemetryProvider(publisher: MyS3Publisher) : TelemetryProvider {
    override val meterProvider = MyMeterProvider(publisher)
    // other members omitted
}

class MyMeterProvider(private val publisher: MyS3Publisher) : MeterProvider {
    override fun getOrCreateMeter(scope: String) = MyMeter(publisher)
}

class MyMeter(private val publisher: MyS3Publisher) : Meter {
    override fun createDoubleHistogram(name: String, units: String?, description: String?) =
        MyDoubleHistogram(publisher, name, units)
    // other members omitted
}

class MyDoubleHistogram(
    private val publisher: MyS3Publisher,
    private val name: String,
    private val units: String?,
) : DoubleHistogram {
    override fun record(value: Double, attributes: Attributes, context: Context?) {
        publisher.put(Measurement(name, units, value, attributes))
    }
}
```

### Final notes

Hopefully these examples will be enough to get you started. There are a few things to note about this post though:

* The examples given here are trivial implementations. They may not be suitable for your needs based on batching or timing requirements, coroutine management, metric metadata, etc. Please adapt them as appropriate for your use case.
* The telemetry API (including metrics) is still marked "experimental" in our API. You must [opt in to using](https://kotlinlang.org/docs/opt-in-requirements.html) our `ExperimentalApi` annotation. We hope to stabilize the API soon and remove this requirement. If you have any feedback as part of your implementation, we'd love to hear it in this discussion or a [feature request](https://github.com/awslabs/aws-sdk-kotlin/issues/new?assignees=&labels=feature-request%2Cneeds-triage&projects=&template=feature_request.yml&title=%28short+issue+description%29).
* The only telemetry provider included with the SDK currently is [the OpenTelemetry provider](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/observability-telemetry-providers-otel.html), which simplifies publishing metrics and other telemetry signals to an OpenTelemetry backend (which itself supports downstream publishing to a variety of destinations). In the future we plan to offer direct integration with CloudWatch ([feature request](https://github.com/awslabs/aws-sdk-kotlin/issues/984)) and X-Ray ([feature request](https://github.com/awslabs/aws-sdk-kotlin/issues/914)).

Let us know if this sufficiently explains metrics publishing support or if there's anything that could be clarified!

<details>
<summary>Full code from the above examples</summary>

This is the full code from the examples above plus no-op implementations for measurements which were unused in this example:

```kotlin
@OptIn(ExperimentalApi::class)
class MyTelemetryProvider(publisher: MyS3Publisher) : TelemetryProvider {
    override val contextManager = ContextManager.None
    override val loggerProvider = LoggerProvider.None
    override val tracerProvider = TracerProvider.None
    override val meterProvider = MyMeterProvider(publisher)
}

class MyMeterProvider(private val publisher: MyS3Publisher) : MeterProvider {
    override fun getOrCreateMeter(scope: String) = MyMeter(publisher)
}

class MyMeter(private val publisher: MyS3Publisher) : Meter {
    override fun createDoubleHistogram(name: String, units: String?, description: String?) =
        MyDoubleHistogram(publisher, name, units)

    override fun createAsyncUpDownCounter(
        name: String,
        callback: LongUpDownCounterCallback,
        units: String?,
        description: String?
    ): AsyncMeasurementHandle = NoOpMeasurementHandle

    override fun createUpDownCounter(name: String, units: String?, description: String?): UpDownCounter =
        NoOpUpDownCounter

    override fun createMonotonicCounter(name: String, units: String?, description: String?): MonotonicCounter =
        NoOpMonotonicCounter

    override fun createLongHistogram(name: String, units: String?, description: String?): LongHistogram =
        NoOpLongHistogram

    override fun createLongGauge(
        name: String,
        callback: LongGaugeCallback,
        units: String?,
        description: String?
    ) = NoOpMeasurementHandle

    override fun createDoubleGauge(
        name: String,
        callback: DoubleGaugeCallback,
        units: String?,
        description: String?
    ) = NoOpMeasurementHandle
}

class MyDoubleHistogram(
    private val publisher: MyS3Publisher,
    private val name: String,
    private val units: String?,
) : DoubleHistogram {
    override fun record(value: Double, attributes: Attributes, context: Context?) {
        publisher.put(Measurement(name, units, value, attributes))
    }
}

object NoOpMeasurementHandle : AsyncMeasurementHandle {
    override fun stop() = Unit
}

object NoOpLongHistogram : LongHistogram {
    override fun record(value: Long, attributes: Attributes, context: Context?) = Unit
}

object NoOpMonotonicCounter : MonotonicCounter {
    override fun add(value: Long, attributes: Attributes, context: Context?) = Unit
}

object NoOpUpDownCounter : UpDownCounter {
    override fun add(value: Long, attributes: Attributes, context: Context?) = Unit
}

data class Measurement(val name: String, val units: String?, val value: Double, val attributes: Attributes)

class MyS3Publisher(private val s3: S3Client, private val coroutineScope: CoroutineScope) {
    private val queue = LinkedBlockingQueue<Measurement>()

    init {
        coroutineScope.launch {
            while (isActive) {
                delay(5.seconds)
                publish()
            }
        }
    }

    fun put(measurement: Measurement) {
        queue.add(measurement)
    }

    private suspend fun publish() {
        val metrics = mutableListOf<Measurement>()
            .also { queue.drainTo(it) }
            .joinToString(separator = "\n")

        s3.putObject {
            bucket = "foo"
            key = "bar"
            body = ByteStream.fromString(metrics)
        }
    }
}
```
</details>

**Reply by RodrasSilva** [reply]
Hello @ianbotsf,

Thank you for your comprehensive response. I've successfully implemented your examples and can now effectively capture the SDK metrics. 🦾

I do have a few inquiries regarding these metrics and their alignment with those published by the Java SDK. In my previous usage of the Java SDK, I routinely monitored the following metrics:

- ServiceCallDuration
- RetryCount
- AvailableConcurrency
- LeasedConcurrency
- Max Concurrency
- PendingConcurrencyAcquires

Upon reviewing the Kotlin SDK metrics documentation, I encountered the [list of exposed metrics](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/observability-telemetry-metrics.html). While I believe ServiceCallDuration corresponds to `smithy.client.call.duration`, I'm uncertain about the equivalence of the others. Could you kindly clarify whether these metrics are indeed exposed by the SDK and, if so, under what names?

Additionally, I'm interested to know if there's a means to ascertain the success or failure of the request that triggered a particular metric (e.g., via checking the status code for a successful response, such as 200).

Looking forward to your insights.

**Reply by ianbotsf** [reply]
These should be equivalent metrics between the SDKs:

| Java SDK metric | Kotlin SDK metric | Notes |
| --- | --- | --- |
| `ServiceCallDuration` | `smithy.client.call.duration` | |
| `RetryCount` | `smithy.client.call.attempts` - 1 | Our metrics track the number of _attempts_ including retries rather than the number of retries directly |
| `AvailableConcurrency` | `smithy.client.http.connections.limit` - `smithy.client.http.connections.usage(state: acquired)` | Computing this metric requires subtracting the number of acquired connections from the total connection limit. |
| `LeasedConcurrency` | `smithy.client.http.requests.usage(state: in-flight)` | |
| `MaxConcurrency` | `smithy.client.http.connections.limit` | |
| `PendingConcurrencyAcquires` | `smithy.client.http.requests.usage(state: queued)` | |

I don't believe there's currently a way to get the specific HTTP response code related to a particular metric. The `smithy.client.call.errors` metric has an attribute `exception.type` which indicates the name of the exception class which was thrown (e.g., `NoSuchBucket`).

**Reply by RodrasSilva** [reply]
Thank you @ianbotsf. I have a few more questions.

### Async Up-Down Counters Setup:
First off, I'm trying to figure out how to set up a meter for async up-down counters. The interface only has the stop method, and I'm not sure how I can retrieve the value in the `createAsyncUpDownCounter` method from the parameters available.

### Grouping Metrics by Request ID:
Also, any idea how I can group metrics by request ID? Like, if one request spits out a bunch of metrics, how do I sort them by request ID? This is helpful so I can calculate the AvailableConcurrency metric (as I need to calculate `smithy.client.http.connections.limit` - `smithy.client.http.connections.usage(state: acquired)`).

### Clarification on Retry Count Calculation:
Also, shouldn't Retry Count be `smithy.client.call.attempts - 1 `instead of `smithy.client.call.attempts + 1` since attempts should be 1 when no restarts occur?

### ETA for Stable Telemetry API:
By the way, is there an ETA of when the telemetry API is stable?

**Reply by ianbotsf** [reply]
More great questions!

### Async up-down counters setup

Async counters and gauges are for measurements that aren't emitted upon the occurrence of any specific SDK event. Take for instance HTTP connection usage: the SDK does not emit a new measurement every time a connection is acquired, released, or closed. Rather, it's up to your provider to determine when to observe the value. You can think of _synchronous_ measurements as "push metrics" and _asynchronous_ measurements as "poll metrics".

The `createAsyncUpDownCounter` method receives a `callback` argument, which is supplied by the SDK and specifies how the measurement will be recorded. When writing your `Meter` implementation, it's up to you to track this callback and invoke it when you want to get a new measurement for the metric. The callback has a parameter for you to pass some object that can receive the value. For example:

```kotlin
class MyMeter(private val publisher: MyS3Publisher, private val coroutineScope: CoroutineScope) : Meter {
    override fun createAsyncUpDownCounter(
        name: String,
        callback: LongUpDownCounterCallback,
        units: String?,
        description: String?
    ) = MyAsyncUpDownCounter(publisher, name, callback, units, coroutineScope)

    // other members omitted
}

class MyAsyncUpDownCounter(
    private val publisher: MyS3Publisher,
    private val name: String,
    private val callback: LongUpDownCounterCallback,
    private val units: String?,
    coroutineScope: CoroutineScope,
) : AsyncMeasurementHandle, LongAsyncMeasurement {
    @Volatile private var isPolling = true

    init {
        coroutineScope.launch {
            // Launch a background coroutine to poll the metric every 5 seconds
            while (isActive && isPolling) {
                delay(5.seconds)
                // callback takes a LongAsyncMeasurement, which MyAsyncUpDownCounter implements
                callback(this@MyAsyncUpDownCounter)
            }
        }
    }

    override fun record(value: Long, attributes: Attributes, context: Context?) {
        publisher.put(Measurement(name, units, value, attributes))
    }

    override fun stop() {
        // Caller is indicating the metric is no longer observable so we should stop polling
        isPolling = false
    }
}
```

A few things to note about this example:
* The `MyAsyncUpDownCounter` class implements both `AsyncMeasurementHandle` (meaning it handles the _lifecycle_ of the metric) _and_ `LongAsyncMeasurement` (meaning it receives the measurement recording call). This is convenient for our trivial example but not required...there may be cases when these implementations should be separate or could be delegated to other shared objects in the code.
* The polling is conducted in a loop with a delay but that's not required either. You may decide to poll during specific events like before/after an operation is invoked or when an exception is thrown.
* The `record` metric invokes the `MyS3Publisher` class we discussed earlier. That example used asynchronous batching to accumulate measurements and then emit them on a 5 second period. Given that my _new_ example shows _measuring_ on a 5 second period, further delaying the metrics publishing downstream may be undesirable.
  
  In this case your publisher may want to offer a different method for callers to send measurements which should be published synchronously. Or you may want to make a common polling loop across _all_ async measurements, pass in a new object on every loop, bundle all the values for a single iteration into a collection, and then publish them synchronously. Use your judgement!

### Everything else

* Connection metrics aren't associated with individual requests, they exist independently of whether a request is executing or not. That's why they're treated as async counters as described in the previous section. Performing math on those metrics sounds like a reason to use a common polling loop across multiple metrics to aggregate the results for a single iteration into a collection/object so that the individual values are grouped for your arithmetic.
  
  For other metrics which _are_ associated with requests (e.g., `smithy.client.call.attempt_duration`), request ID is not an attribute or metadatum we associate. The telemetry design is still in development so that may change.
* Yes, Java's `RetryCount` is equivalent to `smithy.client.call.attempts - 1`, not `+ 1`. That's a silly mistake on my part. 🤦‍♂️
* We do not currently have an ETA for the stabilization of the telemetry API. We're still collecting feedback internally and externally on the design and we'll almost certainly have some additions to make to it. Information on your use cases and things you find easy/hard with the current implementation would certainly be valuable to the process.

**Reply by RodrasSilva** [reply]
Hello again @ianbotsf,

Firstly, I extend a heartfelt thank you for your continuous assistance and prompt responses. Allow me to provide a comprehensive overview of the rationale behind our impending transition and how we currently manage our metric consumption within the legacy SDK.

Currently, I'm working on the migration of an internal project within our company from the Java S3 SDK to the Kotlin S3 SDK. The reason for this transition is primarily the fact that our codebase is predominantly written in Kotlin, and we anticipate leveraging the native coroutine support offered by the Kotlin SDK. The migration process itself has been relatively smooth, notwithstanding a minor hiccup with a reported and subsequently rectified [bug](https://github.com/awslabs/aws-sdk-kotlin/issues/1220).

The challenge now lies in transposing the metrics logic entrenched within the Java SDK. For context, our modus operandi involves consuming metrics disseminated by the SDK and transmitting them to an in-house library, which, in turn, exposes them as Prometheus metrics, with certain naming conventions applied.

In the Java SDK, our approach typically commenced with the instantiation of the S3[Async]Client:

```kotlin
val httpClient = buildHttpClient(config)
val retryPolicy = buildRetryPolicy(config)
val s3configuration = buildS3Configuration(config)

return S3AsyncClient.builder().httpClient(httpClient)
      .region(Region.of(config.region))
      .credentialsProvider { AwsBasicCredentials.create(config.accessKey, config.secretKey) }
      .serviceConfiguration(s3configuration)
     .overrideConfiguration { cfg ->
           cfg.retryPolicy(retryPolicy)
               .addMetricPublisher(S3MetricsPublisher(imageRepositoryMetrics, config.bucketName))
       }
       endpointOverride(URI(config.endpoint)).build()
```
We would pass our own metrics publisher (S3MetricsPublisher) with the following code:
```kotlin
private const val UNDEFINED = "undefined"
private const val API_CALL_UNSUCCESSFUL_STATUS = "503"

private enum class S3Metrics(val key: String) {
    API_CALL_SUCCESSFUL_KEY("ApiCallSuccessful"),
    AVAILABLE_CONCURRENCY_KEY("AvailableConcurrency"),
    LEASED_CONCURRENCY_KEY("LeasedConcurrency"),
    MAX_CONCURRENCY_KEY("MaxConcurrency"),
    PENDING_CONCURRENCY_ACQUIRES_KEY("PendingConcurrencyAcquires"),
    SERVICE_CALL_DURATION_KEY("ServiceCallDuration"),
    STATUS_CODE_KEY("HttpStatusCode"),
    RETRY_COUNT_KEY("RetryCount")
}

private val METRICS_TO_MAP_KEYS = S3Metrics.values().map { it.key }.toSet()

internal class S3MetricsPublisher(
    private val s3Metrics: ImageRepositoryMetrics,
    private val bucket: String
) : MetricPublisher {

    override fun publish(metricCollection: MetricCollection) {
        // Group metrics by key (name) and value (metric value)
        val metrics = metricsToMap(metricCollection)

        metrics[SERVICE_CALL_DURATION_KEY.key]?.let {
            if (metrics[STATUS_CODE_KEY.key] == null && metrics[API_CALL_SUCCESSFUL_KEY.key] == false) {
                s3Metrics.publishServiceCallDuration(bucket, API_CALL_UNSUCCESSFUL_STATUS, it as Double)
            } else {
                s3Metrics.publishServiceCallDuration(
                    bucket,
                    (metrics[STATUS_CODE_KEY.key] ?: UNDEFINED).toString(), it as Double
                )
            }
        }

        metrics[RETRY_COUNT_KEY.key]?.let { s3Metrics.publishRetryCount(bucket, (it as Int).toDouble()) }
        metrics[AVAILABLE_CONCURRENCY_KEY.key]?.let {
            s3Metrics.publishAvailableConcurrency(
                bucket,
                (it as Int).toDouble()
            )
        }
        metrics[LEASED_CONCURRENCY_KEY.key]?.let { s3Metrics.publishLeasedConcurrency(bucket, (it as Int).toDouble()) }
        metrics[MAX_CONCURRENCY_KEY.key]?.let { s3Metrics.publishMaxConcurrency(bucket, (it as Int).toDouble()) }

        metrics[PENDING_CONCURRENCY_ACQUIRES_KEY.key]?.let { concurrentAcquires ->
            s3Metrics.publishPendingConcurrencyAcquires(bucket, (concurrentAcquires as Int).toDouble())
        }
    }

    private fun metricsToMap(metricCollection: MetricCollection): MutableMap<String, Any> {
        val hashMap: MutableMap<String, Any> = HashMap()
        val stack = Stack<MetricCollection>()

        metricsToMap(hashMap, metricCollection)
        stack.addAll(metricCollection.children())
        while (!stack.isEmpty()) {
            val metrics = stack.pop()

            metricsToMap(hashMap, metrics)
            stack.addAll(metrics.children())
        }

        return hashMap
    }

    private fun metricsToMap(hashMap: MutableMap<String, Any>, metrics: MetricCollection) {
        metrics.forEach {
            val key = it.metric().name()

            if (key in METRICS_TO_MAP_KEYS) {
                when (val value = it.value()) {
                    is Duration -> hashMap[key] = value.toMillis().toDouble()
                    else -> hashMap[key] = value
                }
            }
        }
    }

    override fun close() {}

}

```
It's imperative to note that in the Java SDK, the publish method was invoked after each request, thereby furnishing metrics associated with the corresponding operation. This granularity is pivotal for us as it enables the aggregation of multiple metrics pertinent to a singular request, thereby facilitating more nuanced analysis.

The ImageRepositoryMetrics interface serves as the conduit for funneling S3 metrics into our Prometheus metrics repository:
```kotlin
interface ImageRepositoryMetrics {
    fun publishPendingConcurrencyAcquires(bucket: String, value: Double)

    fun publishRetryCount(bucket: String, value: Double)

    fun publishServiceCallDuration(bucket: String, label: String, amt: Double)

    fun publishAvailableConcurrency(bucket: String, value: Double)

    fun publishLeasedConcurrency(bucket: String, value: Double)

    fun publishMaxConcurrency(bucket: String, value: Double)
}
```

We would like to replicate this behavior seamlessly within the Kotlin SDK.

You mentioned that "Connection metrics aren't associated with individual requests." Given this constraint, we're confronted with the challenge of discerning and aggregating metrics about a specific origin or request to enable comprehensive analysis and subsequent publication. We're cognizant of the necessity for such calculations, which mandate the collation and aggregation of metrics to furnish the requisite insights.

In essence, we seek guidance on how to identify and aggregate disparate metrics aligned with a singular origin or request within the Kotlin SDK to facilitate the publication of comprehensive analytical data. Your insights on this matter would be immensely valuable as we navigate this transition.

Let me know if you need any more details that can help with the development of the telemetry API.

Kind regards.

**Reply by ianbotsf** [reply]
Yes, I definitely see the challenges with adapting your prior use of the Java SDK with the current Kotlin SDK telemetry APIs. Let me take your feedback and use case to some representatives from the other SDKs and attempt get some consensus.

For context, one of the reasons for the new telemetry API and its current experimental status is that we're attempting to coalesce the user experience across all AWS SDKs—not just for telemetry but in many other areas too. Whereas in the past it might've been easier to make a decision to change the metrics capabilities for one SDK, now we're trying to stick a common architecture in more areas.

**Reply by RodrasSilva** [reply]
Hello again, @ianbotsf. I hope you are well.
Did you happen to get to present our use cases to the other SDKs representatives? Is this something that different clients would want to be able to migrate from? It would be great to keep this logic in the new SDK.
Thanks 😄

**Reply by ianbotsf** [reply]
Hi @RodrasSilva, I'm working on a POC for supporting a `MetricPublisher`-like interface for the AWS SDK for Kotlin within the framework of the larger telemetry spec. My hunch so far is that it won't actually be too much work or require too much adjustment to the telemetry spec that SDKs are migrating to. I'll let you know when I have something more concrete.

**Reply by RodrasSilva** [reply]
Hello @ianbotsf. I am checking to see if you made any progress on your PoC and if you have anything more concrete I can use for testing. Thank you in advance!

**Reply by ianbotsf** [reply]
Hi @RodrasSilva, just back from an extended vacation. The PoC is still on my todo list and I can let you know when I have a branch ready to testing. Getting more eyes on it will be helpful! 🙂

**Reply by RodrasSilva** [reply]
Hello again, @ianbotsf. I hope your vacations went well. 
Did you have a chance to work on the PoC? Please let me know if you have a branch ready for testing :)
